src/Runtime/Distributions/Automata/Automaton.StateData.cs

```diff
@@ -0,0 +1,152 @@
+            /// <param name="transitions">The outgoing transitions.</param>
+            /// <param name="endWeight">The ending weight of the state.</param>
+            [Construction("GetTransitions", "EndWeight")]
+            public StateData(IEnumerable<Transition> transitions, Weight endWeight)
+                : this()
+            {
+                Argument.CheckIfNotNull(transitions, "transitions");
+
+                this.EndWeight = endWeight;
+
+                foreach (var transition in transitions)
+                {
+                    this.AddTransition(transition);
+                }
+            }
+
+            /// <summary>
+            /// Gets or sets the ending weight of the state.
+            /// </summary>
+            [DataMember]
+            public Weight EndWeight { get; set; }
+
+            /// <summary>
+            /// Gets a value indicating whether the ending weight of this state is greater than zero.
+            /// </summary>
+            public bool CanEnd => !this.EndWeight.IsZero;
+
+            /// <summary>
+            /// Gets the number of outgoing transitions.
+            /// </summary>
+            public int TransitionCount => this.transitionCount;
+
+            /// <summary>
+            /// Creates the copy of the array of outgoing transitions. Used by quoting.
+            /// </summary>
+            /// <returns>The copy of the array of outgoing transitions.</returns>
+            public Transition[] GetTransitions()
+            {
+                var result = new Transition[this.transitionCount];
+                Array.Copy(this.transitions, result, this.transitionCount);
+                return result;
+            }
+
+            /// <summary>
+            /// Adds a transition to the current state.
+            /// </summary>
+            /// <param name="transition">The transition to add.</param>
+            /// <returns>The destination state of the added transition.</returns>
+            public void AddTransition(Transition transition)
+            {
+                if (this.transitionCount == this.transitions.Length)
+                {
+                    var newTransitions = new Transition[this.transitionCount * 2];
+                    Array.Copy(this.transitions, newTransitions, this.transitionCount);
+                    this.transitions = newTransitions;
+                }
+
+                this.transitions[this.transitionCount++] = transition;
+            }
+
+            /// <summary>
+            /// Gets the transition at a specified index.
+            /// </summary>
+            /// <param name="index">The index of the transition.</param>
+            /// <returns>The transition.</returns>
+            public Transition GetTransition(int index)
+            {
+                //Argument.CheckIfInRange(index >= 0 && index < this.transitionCount, "index", "An invalid transition index given.");
+                return this.transitions[index];
```

Comment:
Expression body?

Reply:
I converted Argument.CheckIfInRange into Debug.Assert() and uncommented it, so expression body is not an option anymore